### ISSUE TITLE
Type-Tagged Union Contracts

### ISSUE THREAD

**Issue by Radvendii** [tweag/nickel#1533]
As I was trying to type up an explanation of my union types idea for #1448, I realized that the problem I was trying to solve was not of union **types** (which seems to come down to subtyping rules) but rather union **contracts**. This shouldn't surprise me since that's what Yann said was difficult about union types in the first place.

With that in mind, I want to open a new issue focused on that, and explain my workaround for the problems that make union contracts hard, actually.

To summarize the problem with union contracts, they sometimes need to inspect information that, if we want to stay lazy, we don't have access to. If we try to insert those checks into thunks and backtrack if they fail (a la continuations), we run into big problems.

My proposal is that in cases of ambiguity, we put the onus on the user to specify which variant of the union the value is.

I'll be using the notation `|||` for the union contract, since `|` and `||` are taken in nickel. I think it's ugly though and it's not meant as a serious proposal for the notation.

## Applying a union contract
To determine whether a value matches the contract `Foo ||| Bar`, first we check whether, at a surface level, the value matches `Foo`, `Bar`, both, or neither. For contracts like `Number` or `String`, we can fully check whether they apply. For record contracts, we can check whether the appropriate fields are present (but not whether their values match the contracts for those fields). For functions, we can only check whether the value is, in fact, a function. For flat types we can only check whether they are identical flat types.

If no subcontracts match, the contract doesn't match.

If exactly one matches, the contract matches.

If more than one matches, we're in the interesting / hard case. Without any more information we have to say that the contract **doesn't** match.

So where would this extra information come from?

[Right now](https://github.com/tweag/nickel/blob/3df013169b37d71b0a2693f81735b21a813d6287/core/src/eval/mod.rs#L694-L711),  we forget annotations at runtime. A simple version of what I'm proposing would involve keeping contract annotations around during runtime, attached to the values they were applied to. When a union contract encounters that "multiple subcontracts might match" case, it inspects the value's anntotations, and... here there are some decisions to make, but a simple way to do it is just to look at the last contract applied, and if that matches one of the subcontracts, then we say the value passes the contract.

### Examples

```nickel
# just works
5 | Number ||| String
"hello" | Number ||| String

# fine. only one of the contracts could possibly match
{ foo = 5, bar = 5 } | { foo | Number, bar | Number } ||| { foo | String }

# error: couldn't determine which subcontract to apply.
{ foo = 5, bar = 5 } | { foo | Number, bar | Number } ||| { foo | String, bar | String }

# fine. we can see from the earlier contract application which one should apply
{ foo = 5, bar = 5 } | { foo | Number, bar | Number } | { foo | Number, bar | Number } ||| { foo | String, bar | String }

# realistically, we would wrap this up something like this
let MyContract = {
  Left = { foo | Number, bar | Number },
  Right = { foo | String, bar | String },
  C = Left ||| Right
} in
{ foo = 5, bar = 5 } | MyContract.Left | MyContract.C

# this would also work. Where the discriminating contract and the union contract are not applied in the same place
{
  field | MyContract.C
} & {
  field = { foo = "hello", bar = "world" } | MyContract.Right
}
```

## Destructing unions

How would we destruct these unions? Well, we don't actually have to add anything. There's enough power and information already in nickel to destruct them. For instance,

```nickel
let x | MyContract.C = ... in
if std.is_string x.foo then ... else ...
```

In fact, much of the time people will be using methods like this anyways, because much of the time there will be a separate tag / discriminant field to match on.

Remember also that this is a union type, not an ADT. If you make a union of overlapping types `std.number.PosNat ||| Number`, it will be impossible to destruct those separately, because every `PosNat` is a `Number`.

### A proper `match`

It might be useful to add another match construct that allows us to match on contracts (with flow-typing). 

One way is to try to apply the first contract, and if it applies continue as though we'd applied that contract to that value, and if it doesn't try the next branch. This is dangerous because you either partially apply a contract that will blame later on, or you have to fully evaluate the argument and lose laziness. This is the same problem that makes union contracts hard. Maybe it's not a problem to lose laziness though, since we can always fall back on the previous ugly method.

Another option is to once again use the remembered annotations that we used when we applied the union contract. As long as we're destructing a union, we should always have this information (otherwise the union contract would have failed!) However, this runs the risk of turning into ADTs, because we could do:

```nickel
# flat types C1 and C2 are not equal to each other.
let C1 = fun label value => std.contract.apply Number label value in
let C2 = fun label value => std.contract.apply Number label value in
let x = 5 | C1 | C1 ||| C2 in
# because we remember that we came from `C1`, we can match on it even though this has no structural content.
x |> match {
  C1 => ...,
  C2 => ...,
}
```

### By Convention

Another option is to make it ideomatic to include in the `MyContract` record a `discriminant` function that calculates the discriminant and returns a nice enum that you can match on. At least for unions that contain enough information in themselves to discriminate (as opposed to ones that need a tag).

```nickel
let MyContract = {
  Left = { foo | Number, bar | Number },
  Right = { foo | String, bar | String },
  C = Left ||| Right
  discriminant | C -> [| 'Left, 'Right |]
               = fun x => if std.is_string x.foo then 'Left else 'Right
}
```

and you'd use it like

```nickel
let x | MyContract.C = ... in
x |> match MyContract.discriminant x {
  'Left => ...,
  'Right => ...,
}
```

## As a type

As I mentioned above, having a union **type** seems to be a question of how the subtyping works. Presumably, [like `Dyn`](https://github.com/tweag/nickel/blob/master/rfcs/004-typechecking.md#dyn-inference) we wouldn't want to infer a union type except as a result of an explicit annotation, but any type `T` should be able to upcast to `T ||| Foo`  for any `Foo`. So
```nickel
let if_then_else : forall a. Bool -> a -> a -> a = fun b x y => if b then x else y in
f true 5 "string"
```
Would fail, but if you replaced `5` with `(5 : Number ||| String)` it would succeed, just as it would if you replaced it with `(5: Dyn)`.

As for destructing, `%typeof%` isn't detailed enough. We would need something that can distinguish between different kinds of enums, records, etc. and we would probably want flow-typing.

**Comment by Radvendii** [comment]
Okay, so there's two problems that @vkleen and I discussed. I'll try to present the problems as well as my proposed solutions below.

**Problem 1**

We need to compare contracts for equality. So that in `foo | C1 | (C1 ||| C2)` we can tell that the two `C1`s are the same, and know that the value matches the union type. It turns out that with the way contracts are currently implemented this is not so trivial, and maybe impossible. Contracts are just identified with the term written out in the code. So
```nickel
let Foo = C1 in
foo | Foo | (C1 ||| C2)
```
Would **not** see that `Foo` and `C1` are the same. This obviously breaks referential transparency. If we try to evaluate the term, then we're very sensitive to the way evaluated terms are stored (if we use De Bruijn indices, alpha renaming doesn't lead to different contracts, but otherwise it might).

**Solution 1**

My proposed solution is something we've discussed in the past, which is an opaque contract type. There would be functions for creating a contact from a function on labels and values, or from a record, etc. Maybe these would be inserted automatically in contract position (after `|`).

The advantage of an opaque contract type is that each one can be generated with a unique ID. That way we can track whether two contracts actually come from the same place. Kind of like comparing object pointers, rather than the content. This matches the way I was imagining union types working, where variable assignment (`let Foo = C1` above) wouldn't affect things, but declaring the same contract twice would register as different contracts.

**Problem 2**
If we add any mechanism for matching on the variants (lets say for now only in typed code), we end up with a problem:

```nickel
let foo | { a | Foo} | ({a | Foo } ||| { a | Bar}) = ... in
let f : { a | Foo } ||| { a | Bar } -> Blah
      = fun x => x |> match {
        { a | Foo } => ...,
        { a | Bar } => ...
      }
`aaaaa`

At no point in the above code do we actually check that `foo` has the type `{ a | Foo }`. We never try to access `a`, so we never actually apply that part of the contract. Yet, `f` dispatches based on that information.

**Solution 2**

I looked into how [typescript solves this with flow-typing](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) and thought about how we might apply it in nickel. The basic idea is that as more information becomes known about an object, it's type gets refined. This is the code I'm imagining for nickel:
```nickel
let TaggedUnion = { tag : [| 'Left |], value : Left } ||| { tag : [| 'Right |], value : Right } in
let f : TaggedUnion => Foo
       = fun x => x.tag |> match {
         'Left => # x.value has type Left here
         'Right => # x.value has type Right here
       }
```

So how does this work? A couple things are going on here.

- You can perform actions on a union that would make sense for all its subtypes. So `x.tag` is allowed, because `x` has a field `tag regardless of which subtype it is.
- When you access a component that has different types in different subtypes, it unions the possible types. So here `x.tag` has type `[| 'Left |] ||| [| 'Right|]`
- Somehow `[| 'Left |] ||| [| 'Right |]` is made equivalent to `[| 'Left, 'Right |]`. Morally, those types are equivalent. I'm not sure if they can be equated in all contexts, that requires more thought. But at the very least we should be able to `match` on it like it is.
- They type of `x : Foo ||| Bar` gets transformed in the following ways:
  - `if x == (a : Foo) then <e1> else <e2>`: `x` has type `Foo` in `<e1>`.
  - `if %typeof% x == 'Foo then <e1> else <e2>`: `x` has type `Foo` in `<e1>`, and has type `Bar` in `<e2>`
  - Similarly for `match` rather than `if`
  - If one of the components of a union of records is revelealed to have a more specific type, we reevaluate which of the record subtypes is still possible (it might still end up being a union type, but with fewer subtypes)

**Alternative Solutions to 2**

There are a couple solutions I don't think are very good:

1. Decide this isn't a problem. Part of how lazy contracts work is that if you never try to access the part of it that  doesn't actually apply, you'll never have an error. This isn't satisfying because we **are** dispatching on that information, so it feels like it should be checked at that point.

2. Don't do anything to support matching on union terms. This would require the programmer to enter a bunch of contract annotations everywhere to convince the typechecker things are fine.. I think this is a bad long-term solution, but might be a good first step since it technically works and solving the problem down the road won't break existing code. Much like how a lot of nickel code used to require `| Dyn` annotations.

**Comment by yannham** [comment]
At last, another question I have is around the specifics of the implementation of "reading annotations". Currently annotations are just wrapping nodes, and some of them get erased. We could indeed fix that and  keep annotations around. Then we would have to say when we care about them (otherwise, they should be transparent for evaluation). And then to think on how do they interact with evaluation: for example would `(fun x => x) (1 | PosNat)` still know that it is a `PosNat`? Presumably, union values would be returned with a tag and passed around, and we might want to preserve this information.

It feels like this would take us back to the old metadata system, where we have to carry and propagate annotations around everywhere during evaluation, or we would be losing information. Or maybe this could be stored separately for all values, instead of being a specific node of the AST?

**Comment by Radvendii** [comment]
> If you're trying to solve both problems by proposing a solution to union contracts which would enable union types, and thus to encode ADT, then it's crucial to look at pattern matching indeed.

Yes, this one.

>  The whole point, in the general case, is that you can't just test "if a contract match"

I probably should have used a different word. What I mean here is to test whether it matches "at first glance". What I describe just before that: records having the right fields, Is it a function at all, etc.

> If we want to make [more complicated things] work with flow-sensitive typing, we have to make it so that the discriminant determined by the interpreter for the union has an equivalent predicate test that is also understood by the typechecker.

Yeah, this is why I wrote up `Problem 2`.

> We already have a contract equality checking

Oh! Problem solved then. Never mind.

> However, I don't think [flow typing] solves the very example that you used when defining Problem 2:... How do you solve the issue of dispatching without having any information on the value of a (preferably without forcing a)?

You're right. I shouldn't have switched examples there. The original example can't be done, in my mind. It's a little like taking a union in C code and asking how to dynamic dispatch on the type of the union. With an untagged union you need **some** discriminant.

But the problem still stands. If we take the second example, but use a `match` statement on the type rather than match on tag + flow typing:

```
let L = { tag : [| 'Left |], value : Left } in
let R = { tag : [| 'Right |], value : Right } in
let TaggedUnion = L ||| R in
let f : TaggedUnion => Foo
       = fun x => x |> match_type {
         L => # x has type L here
         R => # x has type R here
       }
let x = ... | L | TaggedUnion in
f x
```

We're still dynamic dispatching on something we never checked.

> the implementation of "reading annotations"

I was imagining values would have an associated "last contract applied" somewhere, whether that's on the node itself or in a table elsewhere. This would be preserved by `fun x => x`, because the original value is returned unchanged (with it's "last contract applied" information). But any operation e.g. `+ 1` would not preserve that information. I haven't really thought through the implications of this, like do we end up with a ton of superfluous information bogging us down.

I wasn't around for the old metadata system, so I don't know what pitfalls I might be dragging us back into.

> Presumably, union values would be returned with a tag and passed around, and we might want to preserve this information.

Uhh... I think we actually don't want to preserve that tag. This was the gist of `Problem 2`. Any way that we expose that information to the nickel code allows dynamic dispatch on information that hasn't been verified.

**Comment by Radvendii** [comment]
Another problem came up while discussing with Viktor and Joe last week

## Problem 3

For values with a tag, do we end up needing to "tag things twice"? Consider
```
# pretend we can assign types to variables
let Left = { tag | [| 'Left |], value | LeftValue } in
let Right = { tag | [| 'Right |], value | RightValue } in
let TaggedUnion = Left ||| Right in
let foo = { tag = 'Left, value = blah } | Left | TaggedUnion in
...
```
It's kind of painful to require that `| Left`, since we already have `tag = 'Left`.

## Discussion of 3

I haven't thought of a solution that I'm happy with, but here are my thoughts

Originally, I thought this would be solved by "at a glance" checking the contracts. In my mind laziness meant that we can't force any thunks, but in this case `tag = 'Left` is being assigned a concrete value, which doesn't need to be computed So we can check whether it fits `[| 'Left |]` or `[| 'Right |]`. When applying `| TaggedUnion`.

But if contract checking **itself** must be lazy (in other words, applying contracts should be delayed until the last minute, not just evaluating values), then this doesn't work.

One option is something like
```
let Left = { tag | [| 'Left |] = 'Left, value : LeftValue } in
let Right = { tag | [| 'Right |] = 'Right, value | RightValue } in
let TaggedUnion = Left ||| Right in
let foo = { value = blah } | Left | TaggedUnion in
...
```
But I'm not sure if I like this. I think contracts affecting the value like this is rather confusing/surprising/hard to reason about.

**Comment by yannham** [comment]
Thanks for the answers. I think I start to have a clearer view now, although there are still details to be fleshed out. My current understanding is the following: what you're proposing is basically to have untagged union types. Union contracts break lazyness, as mentioned multiple times and detailed in the related paper. To make union contracts work, you propose to combine two mechanisms:

- Trying to automatically derive a predicate to discriminate between branches of the union that wouldn't affect lazyness. Typically, this is easy for things such as `String \/ Number`. On such unions, we're good. This isn't so obvious for more structure types, but whatever, we can formalize `derive_predicate(left: Type, right: Type) -> Option<NickelPredicate>` that is incomplete but correct (and make it better over time, providing results for more and more case). That's basically what's proposed at the end of the paper, and the companion blogpost, and what we planned to implement at least for contracts alone.
- When we can't obviously distinguish between union branches, use annotations as a way of tagging.
- If nothing succeeds, the contract/typechecking fails.

Note that although the syntax is different, using an annotation is very much like using a tag. The difference is that the tag here wouldn't really be part of the value representation, but rather a separate metadata, as opposed to usual algebraic datatypes. And it wouldn't always be here.

At first glance, I believe that Problem 2 is fundamentally impossible to solve, in this context. It seems like the whole point of using annotations is to be able to give hint for union branches that can't be discriminated by a non-forcing predicate. Thus, by definition, there's no way the interpreter can check that the annotation is right without actually evaluating something (otherwise, the annotation wouldn't be needed). If the value is ever used, and the value is wrong, then the contract would blame. It might be ok to say that if the value is never used, and you give the wrong tag, then we can't do much (you can already write blatantly false expression that never blames as long as they're not evaluated - it's hard to do much better while preserving lazyness, by definition).

> Uhh... I think we actually don't want to preserve that tag. This was the gist of Problem 2. Any way that we expose that information to the nickel code allows dynamic dispatch on information that hasn't been verified.

I do feel like you do want annotations to propagate like a tag. If at some point the user provides an explicit hint as to which branch of the union a value belongs to, this information should be propagated as much as possible. I don't think there's any value in not doing so (then, if we have a discriminating predicate, we can decide that in this case we ignore the annotation for dispatch - but when we don't, it would just force the user to write the same annotation many time).

**Comment by yannham** [comment]
(I wrote my previous message before "Problem 3" was posted)

As per 3, the thing is that something being a concrete value is a very fragile property, that shouldn't usually be observable by code. In particular, it breaks referential transparency: `{tag = 'Right, value ={}}` would be ok but `let tag' = 'Right in {tag = tag', value = {}}` wouldn't be. Even more, it depends on the surrounding code: `let x = complex_computation_of_tag in {tag = x, value = {}}` wouldn't be dispatchable, but `let x = complex_computation_of_tag in std.seq x {tag = x, value = {}}` might (depending on if you allow to follow variable links to decide if something is a value or not).

Also, program transformations might add indirections. In general, I think we should avoid basing our semantics on something that depends on "is `x` an evaluated thunk already". Contract equality is also not ideal, because it's undecidable in the general case (if you really want to equate contracts that are "equivalent" - if you just want to equate stuff that points to the same definition, it's probably ok though).

Thinking about this more, the very general idea is that, in order to have a well-behaved union on which we can match, we need an oracle `which_branch_of_union : Dyn -> [| 'Left, 'Right |]`. For tagged union, this is obvious: just read the tag. Here we propose to use annotations and derive it automatically sometimes. I think you also mentioned at some point another possibility, which is to let user define this discriminant - which could then force stuff, and that's fine, as long as it's explicitly opt-in by the user -. I wonder how we could articulate all of this together.

Still, at this point, it looks like tagged unions - at least for the statically typed part - are much simpler. We can still implement those ideas for contracts alone, to provide various `or`/`union` operators.

**Comment by Radvendii** [comment]
Yes. That was a very eloquent explanation of what I had in mind.

I also agree with your thoughts on problem 2. It is impossible to "solve", but we can choose to either 
1. decide it's not a problem ("It might be ok to say that if the value is never used, and you give the wrong tag, then we can't do much")
2. not have any mechanism for matching in this way, and only use flow-typing.

> I do feel like you want annotations to propagate like a tag. If at some point the user provides an explicit hint as to which branch of the union a value belongs to, this information should be propagated as much as possible.

Ah, yes. I agree. I was thinking about **after** the union contract is applied, the information about which branch it came from is forgotten. Unless we decide on choice (1) above, in which case we would preserve that information.

**Comment by Radvendii** [comment]
> at this point, it looks like tagged unions - at least for the statically typed part - are much simpler.

Yeah. As I worked out the details it became clear that this was more complicated than I originally anticipated.

I still have some reservations about tagged unions, but that's a separate conversation.

**Comment by yannham** [comment]
Given the discussion, and the fact that ADTs now landed in 1.5, I think we won't pursue this investigation in a foreseeable future for the typed part.
